### DISCUSSION TITLE
columnTransformer for column in newColumns not used?

### DISCUSSION THREAD

**Discussion by sqcsabbey** [simolus3/drift#2445] [Q&A]
With this migration code:

```
        if (from < 22) {
          await m.alterTable(TableMigration(
            segments,
            newColumns: [segments.pointsPrefixCharCode],
            columnTransformer: {
              segments.pointsPrefixCharCode:
                  // random A-Z
                  const CustomExpression('abs(random() % 26) + 65'),
            },
          ));
        }
```

I'm getting the error:

```
I/flutter (27190): 11:42:26.604 (+0:00:00.920339)
I/flutter (27190): 💡 Db migrator 20 to 22
I/flutter (27190): Drift: Sent PRAGMA foreign_keys with args []
I/flutter (27190): Drift: Sent PRAGMA legacy_alter_table with args []
I/flutter (27190): Drift: Sent BEGIN TRANSACTION with args []
I/flutter (27190): Drift: Sent SELECT type, name, sql FROM sqlite_master WHERE tbl_name = ?; with args [segments]
I/flutter (27190): Drift: Sent CREATE TABLE IF NOT EXISTS "tmp_for_copy_segments" ("id" BLOB NOT NULL, "created_at" INTEGER NOT NULL, "updated_at" INTEGER NULL, "feature_id" BLOB NOT NULL REFERENCES features (id) ON DELETE CASCADE, "name" TEXT NULL, "points_prefix" TEXT NULL, "points_prefix_char_code" INTEGER NOT NULL
I/flutter (27190): , "color" INTEGER NOT NULL, PRIMARY KEY ("id")); with args []
I/flutter (27190): Drift: Sent INSERT INTO tmp_for_copy_segments ("id", "created_at", "updated_at", "feature_id", "name", "points_prefix_char_code", "color") SELECT "id", "created_at", "updated_at", "feature_id", "name", "points_prefix_char_code", "color" FROM "segments"; with args []
I/flutter (27190): Drift: Sent ROLLBACK TRANSACTION with args []
I/flutter (27190): 11:42:26.750 (+0:00:01.066276)
I/flutter (27190): ⚠️ Loading table types failed w/ SqliteException(1): while executing, no such column: points_prefix_char_code, SQL logic error (code 1)
I/flutter (27190): ⚠️   Causing statement: INSERT INTO tmp_for_copy_segments ("id", "created_at", "updated_at", "feature_id", "name", "points_prefix_char_code", "color") SELECT "id", "created_at", "updated_at", "feature_id", "name", "points_prefix_char_code", "color" FROM "segments";, parameters: 
```

... it looks like Drift thinks the column was already there and so ignores the columnTransformer? 

---

Edit: I think I figured out what's going on; the database version in this case is going from 20 to 22, and here's the migration for version 21:

```
        if (from < 21) {
          // data population for internal users; getting something in there
          await m.alterTable(TableMigration(
            segments,
            newColumns: [segments.pointsPrefix],
          ));
        }
```

... so, for the version 21 migration, Drift tries to copy the data over through a temp table, and it's expecting a column that gets added in version 22 to be there. 

Is this something that could get fixed in Drift? It looks to me like Drift is looking at the current definition in the code when creating these copy statements, and parsing the current schema from the database file itself seems more appropriate here?

**Comment by sqcsabbey** [comment]
I didn't need that alterTable call for the version 21 migration; using addColumn there instead got this working for me, but 2 alterTable statements in subsequent versions, and having a user upgrade multiple versions at a time, sounds like a valid use-case?

**Comment by simolus3** [comment]
> Is this something that could get fixed in Drift?

At runtime, drift only sees the latest snapshot of the generated code. There are CLI tools to export snapshots of your schema as a JSON and to later generate code representing older schema versions, as used in [this example](https://github.com/simolus3/drift/tree/develop/examples/migrations_example) which is able to refer to older versions like this:

https://github.com/simolus3/drift/blob/01771758789070c9deb8401fe72b925a7087dba6/examples/migrations_example/lib/database.dart#L29-L40

I'm currently thinking of a CLI tool that would generate older schema versions into a more compact form which would be more suitable for this. But you can already use these tools to generate different versions of your schemas.
